core/shared/src/main/scala/bobcats/Algorithm.scala

```diff
@@ -93,3 +97,189 @@ object HmacAlgorithm {
     private[bobcats] override def minimumKeyLength: Int = 64
   }
 }
+
+sealed trait CipherParams
+sealed trait CipherAlgorithm[P <: CipherParams] extends Algorithm
```

Comment:
I'm not fond of this encoding; I think we should move to a more path dependent encoding, especially if we're stamping the algorithm on the key. It will also make it a bit more ergonomic I think.

Only problem is that when I tried encoding it, I got exhaustivity warnings all over the place.